### PR TITLE
feat: add ability to add a list of projects and workflows in favorite #2484

### LINKED ISSUE #2484: feat: as a user, i want to mark a project as starred
As a user, i am member of many groups and projects while I only need a quick access on a few ones.

The feature MUST include: 
* a "star" toggle button on a project home (route `/ui/project/<KEY>`)
* a navbar drop-down menu embedding all projects i starred

The feature SHOULD include:
* a counter on project page that claims how many times a given project has been starred
* a panel on the home page showing a build timeline of my starred projects

The feature MAY include:
* a workflow/application declination, as I am sometimes interested in a single application inside a big project

### PR DESCRIPTION
**Pull request by bnjjj**
1. Description
as a user, i want to mark a project as starred
1. Related issues
close #2484
1. About tests
1. Mentions
cc @FabienM 

@ovh/cds

### CODE REVIEW

File: engine/api/navbar/navbar.go [hunk 1]
```diff
@@ -14,95 +12,195 @@ import (
 )
 
 // LoadNavbarData returns just the needed data for the ui navbar
-func LoadNavbarData(db gorp.SqlExecutor, store cache.Store, u *sdk.User) (data sdk.NavbarData, err error) {
-	var query string
-	var args []interface{}
+func LoadNavbarData(db gorp.SqlExecutor, store cache.Store, u *sdk.User) (data []sdk.NavbarProjectData, err error) {
+	// var query string
```
Comment (yesnault) [inline engine/api/navbar/navbar.go]: delete line

File: engine/api/navbar/navbar.go [hunk 2]
```diff
@@ -14,95 +12,195 @@ import (
...
+	// var args []interface{}
```
Comment (yesnault) [inline engine/api/navbar/navbar.go]: delete line

File: engine/sql/087_favorites.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+-- +migrate Up
+CREATE TABLE IF NOT EXISTS project_favorite (user_id BIGINT, project_id BIGINT, PRIMARY KEY(user_id, project_id));
```
Comment (yesnault) [inline engine/sql/087_favorites.sql]: on delete cascade

File: engine/sql/087_favorites.sql [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+SELECT create_foreign_key('FK_PROJECT_FAVORITE_APPLICATION', 'project_favorite', 'user', 'user_id', 'id');
+SELECT create_foreign_key('FK_PROJECT_FAVORITE_GROUP', 'project_favorite', 'project', 'project_id', 'id');
+
+CREATE TABLE IF NOT EXISTS workflow_favorite (user_id BIGINT, workflow_id BIGINT, PRIMARY KEY(user_id, workflow_id));
```
Comment (yesnault) [inline engine/sql/087_favorites.sql]: on delete cascade

File: ui/src/assets/i18n/fr.json [hunk 5]
```diff
@@ -223,6 +223,10 @@
   "common_end": "Fin",
   "common_error": "Erreur",
   "common_loading_projects": "Chargement des projets...",
+  "common_favorites_updated": "Favoris mis à jour avec succès",
+  "common_favorites_delete": "Supprimer des favoris",
+  "common_no_favorites": "Vous n'avez pas encore ajouter de projets ou de workflows dans vos favoris.",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json]: ajouté

File: engine/api/navbar/navbar.go [hunk 6]
```diff
@@ -14,95 +12,195 @@ import (
...
+	for i, g := range u.Groups {
+		if i == 0 {
+			groupID = fmt.Sprintf("%d", g.ID)
+		} else {
+			groupID += "," + fmt.Sprintf("%d", g.ID)
+		}
 	}
 
-	rows, err := db.Query(query, args...)
+	rows, err := db.Query(query, u.ID, groupID, group.SharedInfraGroup.ID)
```
Comment (fsamin) [inline engine/api/navbar/navbar.go]: why not use gorp select ?

File: engine/api/project/dao.go [hunk 7]
```diff
@@ -294,6 +294,7 @@ var LoadOptions = struct {
 	WithClearKeys                  LoadOptionFunc
 	WithPlatforms                  LoadOptionFunc
 	WithClearPlatforms             LoadOptionFunc
+	WithFavorite                   LoadOptionFunc
```
Comment (fsamin) [inline engine/api/project/dao.go]: favorites

File: engine/api/project/dao.go [hunk 8]
```diff
@@ -420,3 +422,16 @@ func unwrap(db gorp.SqlExecutor, store cache.Store, p *dbProject, u *sdk.User, o
 
 	return &proj, nil
 }
+
+// UpdateFavorite add or delete project from user favorites
+func UpdateFavorite(db gorp.SqlExecutor, projectID int64, u *sdk.User, add bool) error {
```
Comment (fsamin) [inline engine/api/project/dao.go]: flag argument are so ugly :/

File: engine/api/project/dao_dependencies.go [hunk 9]
```diff
@@ -209,4 +209,14 @@ var (
 		proj.Permission = permission.ProjectPermission(proj.Key, u)
 		return nil
 	}
+
+	loadFavorite = func(db gorp.SqlExecutor, store cache.Store, proj *sdk.Project, u *sdk.User) error {
+		var count int64
+		if err := db.QueryRow("SELECT COUNT(1) FROM project_favorite WHERE project_id = $1 AND user_id = $2", proj.ID, u.ID).Scan(&count); err != nil {
```
Comment (fsamin) [inline engine/api/project/dao_dependencies.go]: gorp selectInt

File: engine/api/user.go [hunk 10]
```diff
@@ -155,6 +158,48 @@ func (api *API) getUsersHandler() Handler {
...
+		fmt.Println("params", params)
+
+		switch params.Type {
+		case "workflow":
+			wf, errW := workflow.Load(api.mustDB(), api.Cache, params.ProjectKey, params.WorkflowName, getUser(ctx), workflow.LoadOptions{WithFavorite: true})
+			if errW != nil {
+				return sdk.WrapError(errW, "postUserFavoriteHandler> Cannot load workflow %s/%s", params.ProjectKey, params.WorkflowName)
+			}
+
+			fmt.Println("todelete", wf.Favorite)
```
Comment (fsamin) [inline engine/api/user.go]: remove fmt.println

File: engine/api/user.go [hunk 11]
```diff
@@ -155,6 +158,48 @@ func (api *API) getUsersHandler() Handler {
...
 
+// postUserFavoriteHandler post favorite user for workflow or project
+func (api *API) postUserFavoriteHandler() Handler {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+		params := sdk.FavoriteParams{}
+		if err := UnmarshalBody(r, &params); err != nil {
+			return err
+		}
+
+		fmt.Println("params", params)
```
Comment (fsamin) [inline engine/api/user.go]: remove println

File: engine/api/user/favorite.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+	"database/sql"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/sdk"
+)
+
+// LoadFavorite return favorites from user given its user id
+func LoadFavorite(db gorp.SqlExecutor, userID int64) (sdk.Favorite, error) {
+	var favorite sdk.Favorite
```
Comment (fsamin) [inline engine/api/user/favorite.go]: use grop to avoid boilerplate scanning code

File: engine/api/user/favorite.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"database/sql"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/sdk"
+)
+
+// LoadFavorite return favorites from user given its user id
+func LoadFavorite(db gorp.SqlExecutor, userID int64) (sdk.Favorite, error) {
```
Comment (fsamin) [inline engine/api/user/favorite.go]: please add unit test

File: engine/api/user.go [hunk 14]
```diff
@@ -155,6 +158,48 @@ func (api *API) getUsersHandler() Handler {
 	}
 }
 
+// postUserFavoriteHandler post favorite user for workflow or project
+func (api *API) postUserFavoriteHandler() Handler {
```
Comment (fsamin) [inline engine/api/user.go]: please add unit test

File: ui/src/assets/i18n/fr.json (hunk 5 above)
Comment (fsamin) [reply ui/src/assets/i18n/fr.json]: ajouté

File: engine/api/user.go (hunk 14 above)
Comment (sguiheux) [reply engine/api/user.go]: maybe he does not have this linter : https://github.com/fsamin/nofuncflags

**Comment by fsamin** [comment]
`cdsctl` displays all the favs \o/

**Pull request merged**